### ISSUE TITLE
Questions about status of HDF5 Application

### ISSUE THREAD

**Issue by marandra** [KratosMultiphysics/Kratos#1676]
Hi @jcotela  and @msandre . I have some general questions for you about the current capabilities of the HDF5Application. For example:

- Is it possible to read a mdpa file and store it as HDF5? 
- Is it possible to store result fields? 
- Is it saving HDF5 directly, or through XDMF? Is XDMF supported at all?
- Would it possible to visualize a solution in Paraview then?
- Do you have a representative example case or use instructions? Is configuration a block in ProjectParameters.json?

Thanks!

**Comment by msandre** [comment]
Hi,

- Is it possible to read a mdpa file and store it as HDF5?
You must pass a model part to hdf. So if you create the model part from mdpa file then yes.

- Is it possible to store result fields?
Yes

- Is it saving HDF5 directly, or through XDMF? Is XDMF supported at all?
hdf5 directly. xdmf is supported.

- Would it possible to visualize a solution in Paraview then?
Yes, for this you need a recent version of paraview (5.4 should work).

- Do you have a representative example case or use instructions? Is configuration a block in ProjectParameters.json?
everything works via processes for now. You can see how these define the json. If these don't do what you need, you can create a new process or use the hdf5 model part io directly. The unit tests are the best tutorial for how to do the latter.

**Comment by marandra** [comment]
Thanks @msandre for your prompt response. I'll look at the test cases (and most likely come back with more questions!)

**Comment by jcotela** [comment]
@marandra If I'm not mistaken, currently only nodal results are supported (no integration point data yet). I'll try to get you an use example later, I prepared one some time ago, but I don't have it on this computer.

**Comment by marandra** [comment]
Hi again. I am trying to create/guess an example where an HDF5 file with the model part and nodal results is written at the end of the simulation. I assembled the following block, which writes and RVE.h5 file, but gives and error: `RuntimeError: Error: Invalid path: ./NodalSolutionStep/VariablesList`
I am no sure if it requires a directory structure or if it is referring to groups and datasets structure of the h5 file. Anyways, any feedback will be very welcome. (Also, how should I modify it to write XDMF files?)
Thanks!

```
{
        "python_module": "single_mesh_temporal_output_process",
        "kratos_module": "KratosMultiphysics.HDF5Application",
        "process_name": "HDF5NodalSolutionStepDataIO",
        "Parameters":{
            "model_part_name": "RVE",
            "file_output_settings": {
                "file_access_mode": "exclusive",
                "file_driver": "sec2",
                "echo_level" : 0
	    },
            "model_part_output_settings": {
	        "prefix": "."
	    },
            "results_settings": {
	        "prefix": ".",
		"list_of_variables": ["DISPLACEMENT", "REACTION"]
	    },
            "output_time_settings":{
                "output_time_frequency": 1.0,
                "output_step_frequency": 0,
                "time_tag_precision": 4
	    }
            
        }
}
```

**Comment by msandre** [comment]
Hi @marandra , try removing the dot from the prefix. Only alpha numeric characters and forward slash are allowed in the path name (the path mimics the linux file system directory structure but is local to the hdf5 file). The prefix is like the folder name where you want to store the modelpart / results. E.g., you can make prefix=/MyModelPart for the model part output settings and prefix=/MyNodalResults for results settings. Then you should see the corresponding data writing under these paths in the hdf5 file. Also, your process_name seems invalid but it probably does nothing because of the factory method. I would suggest leaving it empty. The xdmf is a xml file that describes the data structure in the hdf5. Currently it is generated after you have the hdf5 results files written. You simply call the create_xdmf_file.py script on the initial hdf5 file which contains your model part and it should parse all the files in your directory and assemble the xdmf hierarchy.

**Comment by marandra** [comment]
Thanks for the clarifications. I removed the options that I could and the defaults are working fine for me. Now the program is running up to the point where it tries to read the properties, and enters registered_variables_lookup.h. There it fails with
 ```
RuntimeError: Error: Variable "CONSTITUTIVE_LAW" was not found.
Path: "/ModelData/Properties/(1)/DataValues".
```
The properties in my case are:
```
r_data : data value container
    CONSTITUTIVE_LAW : 0x55848a0ea920
    YOUNG_MODULUS : 21000
    POISSON_RATIO : 0.3
    YIELD_STRESS : 5.5
    REFERENCE_HARDENING_MODULUS : 1
    ISOTROPIC_HARDENING_MODULUS : 0.12924
    INFINITY_HARDENING_MODULUS : 0
    HARDENING_EXPONENT : 1
```
I think it isn't related to HDF5Application, as it seems to be missing support for CONSTITUTIVE_LAW variable type in kratos_components.h. I am not really sure how to fix it, so maybe it is better to open a new issue. Do you agree?

**Comment by loumalouomega** [comment]
The problem with the CL is the same in case you use the model_part_io to generate a mdpa. The CL is treated as pointer, so, what it saves is the reference to the CL. I think @philbucher is currently owrking on that

**Comment by jcotela** [comment]
@marandra I think you are right in your assesment, If I'm not mistaken, constitutive law type variables are not yet supported. I think this could be added relatively easy using the new registration of constitutive laws by @philbucher.

**Comment by marandra** [comment]
Ah, ok. It must be related to #1642 or #1654 . So if it is WIP I don't mind waiting a few more days. Otherwise let me know how can I help register the CLs. Thanks!

**Comment by msandre** [comment]
I just spoke to Jordi about this issue. Currently there is no support for CL in hdf5 io. It sounds like the constitutive law is somewhat dependent on element implementation detail. If this is the case, we would need to give the responsibility of serializing the CL to the implementation. It's not clear at the moment (to me at least) how this should be accomplished (e.g., as a json string?). Also how would we access it from the element, through a new public member function? If such decisions are made, one could start thinking about how to design a constitutive_law io for hdf5 and extending the existing io. Probably it is best to make a new issue unless these are considered in the other existing issues you referenced.

**Comment by marandra** [comment]
Hi guys, it is Riccardo here.

my feeling is that we should write to the hdf5 file only the information of the geometries and groups.
all the rest should be handled with the parameters file and the materials.json file.

It would be quite easy if we want to avoid having multiple files to have Parameters and materials written to hdf5 as strings in json format. 

having said this, the case of CLs is particular, since writing down the pointer is good for nothing... we would need to know what the name of the CL is and write down that one instead. This could be accomplished by iteration over the registered CLs until finding a type match and getting the corresponding key in the kratos components. Of course this will not be very useful if the CL has internal data (say a plastic multipler to make an example)

The problem is that for fluids this should not be very difficult (we typically have one single CL), however for structures i am afraid it will immediately get to a very high complexity level.

**Comment by jcotela** [comment]
Well for fluids it is simple... We don't need to serialize the CL data because all of it can also come from somewhere else. It is much more difficult to store the internal data, as you say, and it doesn't help that it is stored "in free format" (the derived CL class defines any data it needs/wants as member variables) and managed by the CL class itself.

The json thing makes a lot of sense but, as always, the fact that string have variable lengths will complicate the things quite a bit for MPI (since everything has to be stored in uniformly-sized table-like structures there, or the different processes just don't know where to write to in the file).

**Comment by marandra** [comment]
It's Marcelo again.

In my particular case, I intend to use HDF5 for visualization, so only geometrical data and some results fields would be enough. I don't intend to use it for restarting, which I think is where things get hairy.  Would it be possible add an option to opt-out for storing material data? Would this be a (partial) solution?

**Comment by jcotela** [comment]
Yes that would be much simpler. Then it is just a matter of what we do with the variable CONSTITUTIVE_LAW in properties (and any other non-standard variable types that we might encounter). @msandre would it be an option to silently skip any "unknown" variable types for now and not write them in the mdpa file?

**Comment by msandre** [comment]
@jcotela I think it will work to write json string as an attribute. This can be done for both mpi and nonmpi by telling each process to write the whole json string.

@marandra I am unsure about why it is writing the CL in the first place. Is it correct that the error happens when reading? I would expect that it should fail when writing the properties. 

@jcotela yes, I think we can do this no problem. I just want to be sure where the problem is. I would still throw from the registered variable lookup and then catch and discard in the WriteDataValueContainer() if indeed the problem is coming from there, which I suspect. There should be no failure to find the variable in the ReadDataValueContainer(). @marandra can you try this on your example?

**Comment by marandra** [comment]
@msandre you are right, it is when writing the properties. This is the complete error message:

```
Traceback (most recent call last):
  File "MainKratos.py", line 92, in <module>
    analysis(parameters, processes, gid_output, solver, model_part)
  File "MainKratos.py", line 24, in analysis
    process.ExecuteBeforeSolutionLoop()
  File "/home/mraschi/Projects/Kratos/applications/HDF5Application/python_scripts/hdf5_output.py", line 184, in ExecuteBeforeSolutionLoop
    self._initial_output.Execute()
  File "/home/mraschi/Projects/Kratos/applications/HDF5Application/python_scripts/hdf5_output.py", line 145, in Execute
    output.Execute(self._model_part, hdf5_file)
  File "/home/mraschi/Projects/Kratos/applications/HDF5Application/python_scripts/hdf5_output.py", line 73, in Execute
    KratosHDF5.HDF5ModelPartIO(hdf5_file, self.settings["prefix"].GetString()).WriteModelPart(model_part)
RuntimeError: Error: Variable "CONSTITUTIVE_LAW" was not found.
Path: "/ModelData/Properties/(1)/DataValues".




in /home/mraschi/Projects/Kratos/applications/HDF5Application/custom_utilities/registered_variable_lookup.h:76:void RegisteredVariableLookup<TVariables>::Execute(Targs& ...) [with TFunctor = WriteVariableFunctor; Targs = {HDF5::File, const __cxx11::string, const DataValueContainer}; TVariables = {Variable<int>, Variable<double>, Variable<boost::numeric::Vector >, Variable<boost::numeric::Matrix >}]
   /home/mraschi/Projects/Kratos/applications/HDF5Application/custom_io/hdf5_data_value_container_io.cpp:60:void HDF5::Internals::WriteDataValueContainer(HDF5::File&, const string&, const DataValueContainer&)
   /home/mraschi/Projects/Kratos/applications/HDF5Application/custom_io/hdf5_properties_io.cpp:47:void HDF5::Internals::WriteProperties(HDF5::File&, const string&, const PropertiesType&)
   /home/mraschi/Projects/Kratos/applications/HDF5Application/custom_io/hdf5_properties_io.cpp:64:void HDF5::Internals::WriteProperties(HDF5::File&, const string&, const PropertiesContainerType&)
   /home/mraschi/Projects/Kratos/applications/HDF5Application/custom_io/hdf5_model_part_io.cpp:210:virtual void HDF5::ModelPartIO::WriteModelPart(ModelPart&)

```

**Comment by marandra** [comment]
Hi @msandre . I was able to generate the XDMF and to open it with Paraview. I had to add  `(3,8): "Hexahedron"` to the topologies structures in xdmf.py (line 116) so it could recognize the hexas of my mesh.

There is one issue is when generating  the xdmf file. The first time I run `python3 create_xdmf_file.py 
RVE.h5`, it complains about Conditions data not existing (which I don't have in my model).
```
  #005: ../../../src/H5Gloc.c line 385 in H5G_loc_find_cb(): object 'Conditions' doesn't exist
    major: Symbol table
    minor: Object not found
Traceback (most recent call last):
  File "/home/mraschi/Projects/Kratos/applications/HDF5Application/python_scripts/create_xdmf_file.py", line 93, in <module>
    main()
  File "/home/mraschi/Projects/Kratos/applications/HDF5Application/python_scripts/create_xdmf_file.py", line 61, in main
    GenerateXdmfConnectivities(file_name)
  File "/home/mraschi/Projects/Kratos/applications/HDF5Application/python_scripts/create_xdmf_file.py", line 12, in GenerateXdmfConnectivities
    KratosHDF5.HDF5XdmfConnectivitiesWriterProcess(file_name, "/ModelData").Execute()
RuntimeError: Error: H5Gopen failed.
```

 However, the HDF5 file is modified, with a new group with XDMF info:
```
GROUP "Xdmf" {
   GROUP "Elements" {
      GROUP "SmallDisplacementBbarElement3D8N" {
         ATTRIBUTE "Dimension" {
            DATATYPE  H5T_STD_I32LE
            DATASPACE  SCALAR
            DATA {
            (0): 3
            }
         }
         ATTRIBUTE "NumberOfNodes" {
            DATATYPE  H5T_STD_I32LE
            DATASPACE  SCALAR
            DATA {
            (0): 8
            }
         }
         ATTRIBUTE "WorkingSpaceDimension" {
            DATATYPE  H5T_STD_I32LE
            DATASPACE  SCALAR
            DATA {
            (0): 3
            }
         }
         DATASET "Connectivities" {
            DATATYPE  H5T_STD_I32LE
            DATASPACE  SIMPLE { ( 48, 8 ) / ( 48, 8 ) }
            DATA {
            (0,0): 48, 43, 73, 77, 70, 67, 85, 90,
            (1,0): 61, 48, 77, 83, 81, 70, 90, 94,
            (2,0): 76, 61, 83, 93, 88, 81, 94, 97,
            ...
```

After that, If a run `python3 create_xdmf_file.py RVE.h5` a *second* time, it successfully generates the RVE.xdmf file, which can be read by Paraview with no problems.

**Comment by msandre** [comment]
thanks @marandra , I'm pretty sure I know what the problem is and will make a pr when I'm finished fixing some other tests (hopefully today).

**Comment by msandre** [comment]
Can we close this?

**Comment by pooyan-dadvand** [comment]
Just a question, is HDF5 process is documented in wiki?

**Comment by msandre** [comment]
At the moment it is not.

**Comment by pooyan-dadvand** [comment]
At least a brief description would be very helpful ...

**Comment by jcotela** [comment]
@pooyan-dadvand I think the "user-level" HDF5 interface is not quite done at the moment. I'd like to have some gid_output style wrapper that can be used more or less interchangeably with the current gid io module, but it still has to be designed. I'd rather write a nice wiki page for that once it is done than to write a quickly obsolete description of the "developer level" interface.

**Comment by pooyan-dadvand** [comment]
I agree!

**Comment by marandra** [comment]
Closing it. Thanks for your help.
